Expanding gym wrapper: Allow for GRAY8, Delta Buttons, Multiple Button Presses

bebeal opened the pull request:
Addressing https://github.com/mwydmuch/ViZDoom/issues/505

1. Allow for GRAY8 screentype
	* works with all combinations of buffers
	* works with rendering
2. Support for DELTA buttons by creating a Dict action space
	* Each delta button is it's own Box action space, with the range set to the delta buttons max value, which if not defined, is set to infinity. 
	* Made it such that delta buttons are always at the front of the one hot encoded action, i.e. [delta_0, ..., delta_n, 0, 1, 0, 0]
	* Each delta button in the Dict with the key "delta_{x}" where {x} is the index of the delta button
	* All of the binary buttons are in one big Discrete/MultiDiscrete space with the key being "binary"
3. Support for multiple buttons pressed at once
	* Change with max_buttons_pressed parameter, default: 1
	* max_buttons_pressed > 1 changes the "binary" action from Discrete -> MultiDiscrete

Here's what some example action spaces would be given the available buttons defined:
* [MOVE_LEFT, MOVE_RIGHT, ATTACK], `max_buttons_pressed=1` == `Discrete(3)`
* [MOVE_LEFT, MOVE_RIGHT, ATTACK], `max_buttons_pressed=2` == `MultiDiscrete([3, 3])`
* [TURN_LEFT_RIGHT_DELTA, MOVE_LEFT, MOVE_RIGHT, ATTACK], `max_buttons_pressed=1` == `Dict("binary": Discrete(3), "delta_0": Box(-inf, inf, (1,), np.float32))`
* [TURN_LEFT_RIGHT_DELTA, MOVE_LEFT, MOVE_RIGHT, ATTACK], `max_buttons_pressed=2` == `Dict("binary": MultiDiscrete([3, 3]), "delta_0": Box(-inf, inf, (1,), np.float32))`
* [TURN_LEFT_RIGHT_DELTA, MOVE_FORWARD_BACKWARD_DELTA] == `Dict("delta_0": Box(-inf, inf, (1,), np.float32), "delta_1": Box(-inf, inf, (1,), np.float32))`

Unsure about this interaction:
Right now `max_buttons_pressed` only takes into account binary buttons and not delta buttons. Meaning if `max_buttons_pressed=2` and the available buttons are [TURN_LEFT_RIGHT_DELTA, LOOK_UP_DOWN_DELTA, MOVE_LEFT, MOVE_RIGHT, ATTACK], the action space is:
`Dict(binary:MultiDiscrete([3 3]), delta_0:Box(-inf, inf, (1,), float32), delta_1:Box(-inf, inf, (1,), float32))` and the user can send a list of only 2 values that one hot encode one of the 3 possible binary actions, but they can still use the delta buttons as well, If they send something like `({"binary":array([1, 0]), "delta_0":0.2, "delta_1":0.3})` which one hot encodes to [0.2, 0.3, 1, 1, 0] resulting in really 4 buttons being "pressed" since the delta buttons complicate things. However I think is acceptable, and is actually the more common use case.

Miffyli commented:
Hey! Thanks for the PR. I am currently swamped with life things, but I will try to get time to go over this sooner rather than later ^^

mwydmuch commented:
Hello @bebeal, thank you very much for the PR and for starting the discussion on the topic, I was quite busy during the last month, but I should be able to check the PR and join the discussion later this week :)

Miffyli commented on gym_wrapper/base_gym_env.py:
```diff
@@ -47,37 +54,80 @@ def __init__(
+        self.delta_buttons = len(delta_buttons)
+        self.binary_buttons = len(binary_buttons)
+
+        if self.delta_buttons == self.binary_buttons == 0:
```
Why this? I do not see a scenario where attack alone would be an useful button to press.

bebeal replied:
Yeah you're right, it's not useful, should I throw an error or not even check ?

Miffyli replied:
Lets throw an error if there are no actions :)

Miffyli commented on gym_wrapper/base_gym_env.py:
```diff
@@ -26,6 +27,8 @@ def __init__(
             level (str): path to the config file to load. Most settings should be set by this config file.
             frame_skip (int): how many frames should be advanced per action. 1 = take action on every frame. Default: 1.
+            max_buttons_pressed (int): how many binary buttons can be pressed at once. 1 results in a discrete action space.
+                                       Otherwise results in a MultiDiscrete action space. Default: 1
```
Hmm the point of this is that you limit the number of Discrete choices: if you do all possible combinations of N buttons, you get a lot of options. Limiting to "M pressed buttons at once" limits the Discrete space size.

MultiDiscrete bypasses this issue completely, and I do not think it should be limited in this way. I would maybe change this to `if max_buttons_pressed is None, then use MultiDiscrete space instead`. This also allows people to use Discrete space with more than one button pressed at once in case they do not have an implementation for MultiDiscrete space.

Clarification: in `MultiDiscrete` space, the idea is that each button has on/off or 0/1 states, and you individually choose the on/off state for each button.

bebeal replied:
Ah I guess I misunderstood how `MultiDiscrete` works. Here was my thought process:

Say there are 4 in-game buttons defined (`ATTACK`, `MOVE_LEFT`, `MOVE_RIGHT`, `MOVE_FORWARD`), and `max_buttons_pressed=2`. All possible combinations of N buttons is 
```
[0, 0, 0, 0] // no op
[1, 0, 0, 0] // ATTACK
[0, 1, 0, 0] // MOVE_LEFT
[0, 0, 1, 0] // MOVE_RIGHT
[0, 0, 0, 1] // MOVE_FORWARD
[1, 1, 0, 0] // ATTACK + MOVE_LEFT
[1, 0, 1, 0] // ATTACK + MOVE_RIGHT
[1, 0, 0, 1] // ATTACK + MOVE_FORWARD
[0, 1, 1, 0] // MOVE_LEFT + MOVE_RIGHT (effectively no_op...)
[0, 1, 0, 1] // MOVE_LEFT + MOVE_FORWARD
[0, 0, 1, 1] // MOVE_RIGHT + MOVE_FORWARD
```
To avoid the action space becoming `Discrete(11)`, which with a more complicated example becomes way too many options for all possible combinations, I thought the users action would instead be a list with $(0 <$`len(action)`$<= 2)$ (could can also allow for just an `int` easily enough) where each value is within $[0, 3]$. An example action could be `[1, 2]` which corresponds to `[0, 1, 1, 0]`. I figured this was `MultiDiscrete([2, 4])`, but I misunderstood how `MultiDiscrete` works. 

Is this more akin to just having the action space be `Discrete(4)` and allowing for a list of ints (or just a single int)? How were you thinking this action space would work? Also are we forcing the user to provide `max_buttons_pressed` number of actions, or just up to `max_buttons_pressed` number of actions.

Miffyli replied:
Ah, fair confusion: `MultiDiscrete` is essentially multiple `Discrete` choices.

If we only had a single button we can press down, then we would have `Discrete(2)` space (two choices) or an equal `MultiDiscrete([2,])` space.

If we have two buttons, both can be pressed down or be up, so both have `Discrete(2)` space. We _could_ do all possible combinations and create the `Discrete(2 * 2)` space, but we can also do `MultiDiscrete([2,2])` space. Essentially, you choose for each button individually if button should be down or not.

For four buttons, this would be `MultiDiscrete([2,2,2,2])`.

I was thinking that if `max_buttons_pressed=None` (or -1 or something special), then you could just represent all binary buttons with `MultiDiscrete`

bebeal replied:
I see.

Final clarification then: 

For the case where `max_buttons_pressed` is positive, would you like the action space to be a `gym.spaces.Tuple` of `Discrete(2)` with `len == max_buttons_pressed`? (And obviously if `max_buttons_pressed==1` we can rid the Tuple, and just have `Discrete(2)`)?

In this case, is the user still sending in a list of indices indicating which binary buttons are "pressed", i.e. `[0, 2]` would create the environment action: `[1, 0, 1, 0]`? (and again they can easily just send an int for a single action, etc, that's no issue)

bebeal replied:
To be precise, something like this:

if `max_buttons_pressed > 0`: binary action space is `Tuple` of `Discrete(2)` with `len == max_buttons_pressed`.
(potentially can have a special case when `max_buttons_pressed == 1`, binary action space is `Discrete(2)`)

if `max_buttons_pressed < 0`: binary action space is `MultiDiscrete([2] * -max_buttons_pressed) `.

Miffyli replied:
Ah, sorry, this is what I had in mind (but I can see where the confusions are coming up; this suggestion is not elegant either 😅 ):

If `max_buttons_pressed > 0`: Create a _single_ `Discrete(n)` space for pressing the binary buttons, where `n = 2 * 2 * 2... = 2^num_buttons` (i.e. all possible combinations of pressing `num_buttons` different buttons).

If `max_buttons_pressed == -1 (or 0)`: Create a single `MultiDiscrete([2, 2, 2, ...]) = MultiDiscrete([2,] * num_buttons)` space, where the press action is decided for each button separately.

Any other `max_buttons_pressed` raises an error which tells of these two options.

Miffyli commented on gym_wrapper/base_gym_env.py:
```diff
@@ -37,7 +40,11 @@ def __init__(
-        Action space is always a Discrete one, one choice for each button (only one button can be pressed down at a time).
+        Action space can be Discrete (only one button can be pressed down at a time), MultiDiscrete (multiple buttons can be pressed), Box (Continuous), or a dictionary of some combination
+        If dictionary, the entires are:
+          "binary"        = Discrete(x)/MultiDiscrete(x) depending on if max_buttons_pressed == 1 or not
```
Update this to reflect updates to `max_buttons_pressed`.

Miffyli commented on gym_wrapper/base_gym_env.py:
```diff
@@ -47,37 +54,80 @@ def __init__(
-        if screen_format != vzd.ScreenFormat.RGB24:
-            warnings.warn(f"Detected screen format {screen_format.name}. Only RGB24 is supported in the Gym wrapper. Forcing RGB24.")
+        if screen_format != vzd.ScreenFormat.RGB24 and screen_format != vzd.ScreenFormat.GRAY8:
+            warnings.warn(f"Detected screen format {screen_format.name}. Only RGB24 and GRAY8 is supported in the Gym wrapper. Forcing RGB24.")
```
```suggestion
            warnings.warn(f"Detected screen format {screen_format.name}. Only RGB24 and GRAY8 are supported in the Gym wrapper. Forcing RGB24.")
```

Miffyli commented on gym_wrapper/base_gym_env.py:
```diff
@@ -130,20 +180,42 @@ def __init__(
+                act[delta_idx] = action["delta_" + str(delta_idx)][0]
+
+    def __build_action(self, action):
+        # convert action to vizdoom action space (one hot)
```
Nitpick, but the action space is not one hot here if it is dictionary or something :). "One hot" referes to a vector with all but one entry being zeros, and the chosen one being one.

bebeal replied:
yeah it's more multi-hot encoded or just the "encoded" action

Miffyli replied:
Maybe just "environment action" (I personally use terms "agent action" [what comes out from agent network] and "environment action" [what goes into step]), but "encoded" works too :)

Miffyli commented on gym_wrapper/base_gym_env.py:
```diff
@@ -199,6 +275,8 @@ def __build_human_render_image(self):
 
         image_list = [game_state.screen_buffer]
+        if self.channels == 1:
+            image_list = [np.repeat(game_state.screen_buffer[..., None], repeats=3, axis=2)]
```
Why repetition? If we only have a single-channel image, we should probably just return a single-channel image instead of replicating it over multiple dimensions.

bebeal replied:
Yeah I guess it doesn't need to, I was just following what was already done with the depth buffer below this, but I'm assuming that had to be done because the rgb and automap were `[H, W, 3]` and thus the depth buffer was made to match?

Miffyli replied:
Hmm depth buffer seems to be without a channel (i.e. a single channel), no stacking involved. For consistency and clarity, you could add the channel dimension to all observations: `depth` buffer would become `[H, W, 1]` instead of just `[H, W]`. `image` and `automap` would also always have channel dimension, but it would vary between 1 and 3 depending on which screen format was chosen.

bebeal replied:
Alright in my local version of this pr I did this (making the default observation `[H, W, 1]` for `depth`, `labels` buffer. Definitely agree this is better for consistency.

But for the rendering component (`__build_human_render_image()`) you do need to repeat the color channels so that there are always 3 `[H, W, 3]` for each buffer for 2 reasons:
1. The current way the renderer is implemented using pygame always expects rgb images `[H, W, 3]` and breaks with a single color channel `[H, W, 1]` 
```
surf = pygame.surfarray.make_surface(render_image)
  File "/home/bebeal/anaconda3/envs/doomDev/lib/python3.8/site-packages/pygame/surfarray.py", line 100, in make_surface
    return pix_make_surface(array)
ValueError: must be a valid 2d or 3d array
```
2. If the labels buffer is active then it's an outlier in that is always an rgb image `[H, W, 3]` regardless of screen type and since we stack the buffers horizontally to render the rest of them have to have the same shape `[H, W, 3]`. 

The repeat hack does work for all combinations of buffers:
![render](https://user-images.githubusercontent.com/42706447/175252548-a59cae74-6470-4208-9c40-53a90e1b6194.png)

bebeal replied:
So I mean we could change it to avoid repeating, but it's a little more involved. I think repeating for now is fine unless you think otherwise.

Miffyli commented on gym_wrapper/base_gym_env.py:
```diff
@@ -207,15 +285,18 @@ def __build_human_render_image(self):
-            image_list.append(game_state.automap_buffer)
+            automap_buffer = game_state.automap_buffer
+            if self.channels == 1:
+                automap_buffer = np.repeat(automap_buffer[..., None], repeats=3, axis=2)
```
Same as above: should probably only return single channel image if that is coming from the game (not repeat).

Miffyli reviewed (commented):
Sorry for the delay! I made bunch of comments, but mostly looks good :)

Could you also update `tests/test_gym_wrapper.py` to test these settings, e.g. create new env config to test these out.

bebeal commented:
Okay addressed many/all? of the feedback with the following changes, and have some questions/concerns we can discuss below. 

Also I re-organized a tiny bit of the existing code regarding line length, and things such as putting the lengthy observation space definitions in a function `__get_observation_space()`. In general I provided comments throughout most of the functionality I impelemented, but feel free to remove/add some, whatever you think is cleanest in the code and provides clearity.

onto the questions/concerns:

1: Handling multiple button presses
I currently implemented this by grouping together the environment actions based on the number of set bits:
The benefit of this approach is we can simplify the action space further from 
`Discrete(num_binary_buttons)` $\rightarrow$ `Discrete(num_valid_actions)` where `num_valid_action`
is defined by the number of environment actions that have $\leq$ `max_button_pressed` number of set 
bits. Example:

Say `num_binary_buttons=4` (referencing [`MOVE_LEFT`, `MOVE_RIGHT`, `TURN_LEFT`, `ATTACK`]).
The various encoded "environment actions", and corresponding "agent actions" consists of:

```
env action   agent action    
[0, 0, 0, 0]        0 
[0, 0, 0, 1]        1
[0, 0, 1, 0]        2
[0, 1, 0, 0]        3
[1, 0, 0, 0]        4
[0, 0, 1, 1]        5
[0, 1, 0, 1]        6
[1, 0, 0, 1]        7
[0, 1, 1, 0]        8
[1, 0, 1, 0]        9
[1, 1, 0, 0]        10
[0, 1, 1, 1]        11
[1, 0, 1, 1]        12
[1, 1, 0, 1]        13
[1, 1, 1, 0]        14
[1, 1, 1, 1]        15
```

We can easily include or not include the no_op, potentially even let the user choose
via another environment parameter, but we can discuss what we think is best.

But assuming we have the no op action, the various action spaces for varying `max_buttons_pressed`
for this example would be:



| `max_buttons_pressed` | binary action space |
| --- | --- |
| $0$ | `MultiDiscrete([2, 2, 2, 2])` | 
| $1$ | `Discrete(5)`  | 
| $2$ | `Discrete(11)` |
| $3$ | `Discrete(15)` |
| $4$ | `Discrete(16)` |


How it changes the action conversion in `step()` is as follows

We already have stored a list of the numerical value for each of the environment action bit strings.
We use this list as a map where the int agent action is used as an index, and then we just encode
the bit string for the corresponding numerical value, and that is what is sent to the environment
as the environment action. Ex:

`self.button_map = [0, 1, 2, 4, 8, 3, 5, 6, 9, 10, 12, 7, 11, 13, 14, 15]`
action: `3` $\rightarrow$ `self.button_map[3]==4` $\rightarrow$ `[0, 1, 0, 0]`

1.1: We definitely can change up this approach, but this seemed to be a much better idea to the alternative which
is always making the action space `Discrete(n)` (where `n` is all combinations of `num_binary_buttons`), and 
directly converting the agent action to it's corresponding bit string and using it as the environment action, 
because this has the odd affect that if `max_button_pressed=1` the valid agent actions are `[0, 1, 2, 4, 8]` 
which is odd since it skips around and there really isn't `n` possible actions there are `5`. And again we
can easily exclude the noop, and have agent action `1` correspond to `[0, 0, 0, 1]`, etc. How is this implementation?

2: Mixing and matching binary/delta buttons

Right now, all possible action spaces:

* Only binary buttons defined, no delta buttons defined:
  * `MultiDiscrete`
  * `Discrete`
* Only delta buttons defined, no binary buttons defined:
  * `Box()`
  * `Tuple(Box(), ...)`
* Both binary and delta buttons defined:
  * `Dict("binary": (MultiDiscrete|Discrete), "continuous", (Box()|Tuple(Box(), ...)))`

Obviously I don't think it's ever really useful to only have delta buttons defined (maybe for a maze solver using just moving?)
but it's simple enough to allow for so I did.

2.1: The question at hand is should we always have the `Dict` action space and then if there are no delta buttons it 
would just have a single key,value pair being "binary" and the binary action space? Or is the current implementation 
preferred? I feel like the most common use case of this wrapper is with only binary buttons thats the only reason why
I get rid of the dictionary if there are no delta buttons defined cause I figured it's easier to use if you don't
have to wrap the agents action in the dictionary, but open to changing it.

2.2: Same thing for the "continuous" action space, should I just always make it `Tuple(Box(), ...)` even if there is only
one delta button i.e., len of the tuple is always 1?


3: Testing

I added some tests to cover:
* observation space testing with both screen types `(GRAY8, RGB24)` for various combinations of buffers `(screen|depth|labels|automap)`
* action space testing for all possible combinations of action spaces + varying `max_buttons_pressed` `range(0, 4)`.

I added a directory with some config files that have the arbitrary naming scheme of `<scenario name>_<screen type>_<(i|d|l|a) indicating buffers>_<num_binary_buttons>_<num_delta_buttons>`. 

But I know when the gym wrapper was implemented originally there was [talk](https://github.com/mwydmuch/ViZDoom/pull/499#issuecomment-1068016076) about custom scenario files used for testing, so we can definitely change this up. Specifically:

3.1: Where to place custom config files and how does path to them work? I'm currently just specyfing their location in the code with this: `os.path.join(./<dir>, <scenario file name>)` which works on my local insstance but idk how this pathing (`./`) will work when the repo is pulled or used if installed via `pip`, etc, so how should this be handeled?

Miffyli commented:
## 1. Actions

Sounds to me we both got onto the same wavelength :). This is more or less what I proposed (where I used "combinations of button presses", you seem to say "bit string"). Lets keep it as it is now: noop is ok to have.

## 2.1 Always dict

Lets keep it as is (if only binary buttons -> `Discrete` or `MultiDiscrete` space). This allows people to use most of the libraries this way, which usually have no support for `Dict` space :)

Miffyli commented on gym_wrapper/base_gym_env.py:
```diff
@@ -47,103 +61,93 @@ def __init__(
+            if isinstance(agent_action, int):
+                # agent action maps to numerically equivalent environment action (i.e. bit string is the env action)
+                agent_action = self.button_map[agent_action]
+                agent_action = np.array([i for i in range(agent_action.bit_length()) if agent_action & (1 << i)])
```
Sidenote: this logic might be difficult to debug if errors arise. You could use something [like this](https://github.com/Miffyli/rl-action-space-shaping/blob/master/envs/vizdoom_environment.py#L151-L171) to create a list of valid actions ("binary strings") and then map the discrete action number to action by indexing that list.

bebeal commented:
> ## 1. Actions
> Sounds to me we both got onto the same wavelength :). This is more or less what I proposed (where I used "combinations of button presses", you seem to say "bit string"). Lets keep it as it is now: noop is ok to have.
> 
> ## 2.1 Always dict
> Lets keep it as is (if only binary buttons -> `Discrete` or `MultiDiscrete` space). This allows people to use most of the libraries this way, which usually have no support for `Dict` space :)
> 
> ## 2.2 Tuple for box
> Tuple is even less supported than Dict. Combine continuous actions into one `Box` space (i.e. the becomes (2,))
> 
> ## 3. Tests
> Cheers! Will run these and check them through. I can also see if the pathing works out :). Generally, though, you either run the python test script in the directory as you would any other script, or pytest/other test framework runs it. Tests are not generally shipped as part of the pypi package.
> 
> PS: Sorry for taking so long with this PR, but soon this is done! So far it seems like only 2.2 needs bit adjustment :)

Haha yeah I think we finally got it! Addressed these things and cleaned up a few other things. 

Yeah I was running the test like I would any other script, but didn't know if y'all had some framework setup that would automatically run them and potentially have a different path, etc. :-)

Miffyli reviewed (approved):
Thanks a ton! Tests pass nicely (as long you use gym==0.23 specifically 🙈 ), and they seem to be very exhaustive! Thanks for the hard work you put into this :)
